language/src/regent/optimize_index_launches.t

```diff
@@ -858,6 +867,17 @@ local function analyze_is_simple_index_expression(cx, node)
     node, true)
 end
 
+-- `p` must be a partition or cross product
+local function analyze_is_loop_invariant_recursive(cx, p)
+  if not analyze_is_loop_invariant(cx, p) then
+    if std.is_cross_product(std.as_read(p.value.expr_type)) then
```

Comment:
Shouldn't this be based on the AST structure not the type?

Reply:
I don't understand what is wrong here. If we have `cp[i][i][i]` in an index launch, then we only need to ensure that the base cross product `cp` is loop invariant. Since `analyze_loop_invariant` returns `false` for intermediate cross products (`cp[i][i]`, for instance), I check whether that is the case and recursively call it on the next cross product one level up in the AST (`cp[i]`, in this case).